### DISCUSSION TITLE
changing the mailing provider.

### DISCUSSION THREAD

**Discussion by kelcheone** [digitalocean/supabase-on-do#33] [Q&A]
I have been trying to find out how to change the mail provider from Sendgrid to  something we are using internally. I'm not a Terraform guy but can't really find out how they plug into each other.

**Comment by timothymamo** [comment]
Hey @kelcheone, as you're using an internal mail provider you would probably have to create your own terraform module (which if you're not confident/familiar with I wouldn't suggest you do).

But if you've already got the mail server up and running, with the email sender created and verified, I'd suggest removing the [SendGrid module](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/sendgrid.tf) and sendgrid variables (lines [72-85](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/variables.tf#L72-L85) and lines [195-211](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/variables.tf#L195-L211)) as well as modify the remaining smtp variables (host/port/password/etc) accordingly to your internal server.

Once you've done that you should be all good to go (assuming I haven't forgotten something).

**Reply by kelcheone** [reply]
I think that is pretty much everything. But what is the purpose of the SendGrid module if we're not using SendGrid yet we are string provisioning it?
```hcl
provider "sendgrid" {
  api_key = var.sendgrid_api
}
```
or is it used somewhere else that I don't understand?

**Reply by timothymamo** [reply]
Ah yes, you can remove the `sendgrid` provider too (both the [`required_provider`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L25-L28) section and the [`provider`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L42-L44) itself).

The way terraform works is with the use of providers (you can find the list of providers being used in the [provider.tf](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf) file under the [`required_providers`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L4-L34) part). This is what terraform downloads when you run the `terraform init` command and stores them within the `.terraform/providers/registry.terraform.io` directory.

Some providers need tokens/api keys/access keys/etc to be used as they need to communicate via the providers api (in our case only the [`digitalocean`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L36) and [`sendgrid`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L42) providers need these. Other providers, like [`time`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L29) and [`jwt`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/provider.tf#L17) don't, as they don't access any api, but are used to compute information/data which is needed to create resources.

Hope this helps.
